web/packages/teleterm/src/ui/services/workspacesService/accessRequestsService.ts

```diff
@@ -54,42 +62,249 @@ export class AccessRequestsService {
+  }: ResourceRequest): Promise<void> {
+    if (!(await this.canUpdateRequest('resource'))) {
+      return;
+    }
     this.setState(draftState => {
-      const kindIds = draftState.pending[kind];
-      if (kindIds[resourceId]) {
-        delete kindIds[resourceId];
+      if (draftState.pending.kind !== 'resource') {
+        draftState.pending = {
+          kind: 'resource',
+          resources: new Map(),
+        };
+      }
+
+      const { resources } = draftState.pending;
+
+      if (resources.has(resource.uri)) {
+        resources.delete(resource.uri);
       } else {
-        kindIds[resourceId] = resourceName || resourceId;
+        // Store only properties required by the type.
+        if (kind === 'app' || kind === 'database' || kind === 'kube') {
+          resources.set(resource.uri, {
+            kind,
+            resource: { uri: resource.uri },
+          });
+        } else {
+          resources.set(resource.uri, {
+            kind,
+            resource: { uri: resource.uri, hostname: resource.hostname },
+          });
+        }
       }
     });
   }
+
+  async addOrRemoveRole(role: string): Promise<void> {
+    if (!(await this.canUpdateRequest('role'))) {
+      return;
+    }
+    this.setState(draftState => {
+      if (draftState.pending.kind !== 'role') {
+        draftState.pending = {
+          kind: 'role',
+          roles: new Set(),
+        };
+      }
+
+      const { roles } = draftState.pending;
+      if (roles.has(role)) {
+        roles.delete(role);
+      } else {
+        roles.add(role);
+      }
+    });
+  }
+
+  /**
+   * Combining role access request and resource access request is not allowed.
+   * If the user already has an item for one group, we need to ask
+   * if they want to clear the request before adding items from another group.
+   */
+  private async canUpdateRequest(
+    newRequestKind: 'resource' | 'role'
+  ): Promise<boolean> {
+    let shouldProceed = true;
+    if (
+      this.getState().pending.kind !== newRequestKind &&
+      this.getAddedItemsCount() > 0
+    ) {
+      shouldProceed = await new Promise(resolve =>
+        this.modalsService.openRegularDialog({
+          kind: 'change-access-request-kind',
+          onCancel: () => resolve(false),
+          onConfirm: () => resolve(true),
+        })
+      );
+    }
+    return shouldProceed;
+  }
 }
 
-export function getEmptyPendingAccessRequest() {
+/** Returns an empty access request. We default to the resource access request. */
+export function getEmptyPendingAccessRequest(): PendingAccessRequest {
   return {
-    node: {},
-    db: {},
-    kube_cluster: {},
-    app: {},
-    role: {},
-    windows_desktop: {},
-    user_group: {},
+    kind: 'resource',
+    resources: new Map(),
   };
 }
+
+export type PendingAccessRequest =
+  | {
+      kind: 'resource';
+      resources: Map<ResourceUri, ResourceRequest>;
+    }
+  | { kind: 'role'; roles: Set<string> };
+
+export type ResourceRequest =
```

Comment:
Why such type? It matches the shape of a resource in the search bar and in the unified resources view, so we will be able to simply call:
```ts
accessRequestsService.addOrRemoveResource(resource);
```

Reply:
I'm not sure if convenience should be the dominant factor when making decisions like this one. I'd focus predominantly on whether it fits the domain and, in this particular case, if we'd ever want to reuse it outside of teleterm.

I think the type is fine though. I feel like the comment you left here on GitHub should be turned into a code comment, so that the next person looking at `ResourceRequest` will know about this.

Reply:
>I'm not sure if convenience should be the dominant factor when making decisions like this one. I'd focus predominantly on whether it fits the domain and, in this particular case, if we'd ever want to reuse it outside of teleterm.

Ah, I didn't write the whole story. At first I had a type like `{kind: 'server', uri: ServerUri, hostname: string} | {kind: 'database', ...}`. Unfortunately, it required additional logic to map the unified view or search bar resource to this type. Then I thought that I could match the shape of the other resources to make my life easier. After that I happily wrote this comment, saying only about the convenience part :)

I left a JSDoc comment too.

Reply:
The fact that it matches the shape of resources in _both_ the search bar and the unified resources view makes it a much more convincing argument for me than just the fact that it coincidentally matches the shape of resources in the search bar.